ocp_resources/user_defined_network.py

```diff
@@ -0,0 +1,483 @@
+                occurred. Default is None
+            wait_timeout (int, optional): The maximum time to wait for the conditions to be satisfied (in seconds).
+            sleep_interval (int, optional): The interval between checks (in seconds).
+
+        Returns:
+            dict: The condition that indicates the desired status when met.
+
+        Raises:
+            WaitForStatusConditionFailed: If any of the unexpected conditions is met.
+            TimeoutExpiredError: If the timeout expires before the conditions are satisfied.
+        """
+        samples = TimeoutSampler(wait_timeout=wait_timeout, sleep=sleep_interval, func=lambda: self.conditions)
+
+        try:
+            for sample in samples:
+                for condition in sample:
+                    if any(wait_fn(condition) for wait_fn in wait_condition_fns):
+                        return condition
+
+                if not_wait_condition_fns:
+                    for condition in sample:
+                        if any(not_wait_fn(condition) for not_wait_fn in not_wait_condition_fns):
+                            raise WaitForStatusConditionFailed(
+                                f"Failed to wait for the intended status for UDN {self.name}. "
+                                f"Condition message: {condition['message']}"
+                            )
+
+        except (TimeoutExpiredError, WaitForStatusConditionFailed) as e:
+            self.logger.error(f"{str(e)}")
+            raise
+
+        return {}
+
+    def wait_for_status_condition_ready(
+        self,
+        wait_timeout: int = 120,
+        sleep_interval: int = 2,
+    ) -> dict:
+        """
+        Wait for the UserDefinedNetwork to reach a ready condition status.
+
+        Args:
+            wait_timeout (int, optional): The maximum time to wait for the condition to be met, in seconds.
+            sleep_interval (int, optional): The time to sleep between status checks, in seconds.
+
+        Returns:
+            dict: The condition that indicates the desired status when met.
+
+        Raises:
+            WaitForStatusConditionFailed: If any of the unexpected conditions are met.
+            TimeoutExpiredError: If the timeout expires before the conditions are satisfied.
+        """
+        return self.wait_for_status_condition(
+            wait_condition_fns=[self.is_ready_condition],
+            not_wait_condition_fns=[self.is_sync_error_condition],
+            wait_timeout=wait_timeout,
+            sleep_interval=sleep_interval,
+        )
+
+
+class TopologyType:
+    """
+    This class contains constants for different network topology types used in the UserDefinedNetwork configuration.
+
+    Attributes:
+        LAYER2 (str): Represents a Layer2 topology.
+        LAYER3 (str): Represents a Layer3 topology.
+        LOCALNET (str): Represents a LocalNet topology.
+    """
+
+    LAYER2 = "Layer2"
+    LAYER3 = "Layer3"
+    LOCALNET = "LocalNet"
```

Comment:
I will remove localnet either way as it will not be supported in 4.17:
 https://github.com/ovn-org/ovn-kubernetes/commit/fc7d24ba042a8cc577358115c3496ced5d80cfdb